[CI:DOCS] Update podman.1.md, image-(diff, mount, exists) & images (#10536)
containers/podman: docs/source/markdown/podman.1.md

% podman(1)

## NAME
podman - simple management tool for pods, containers and images

## SYNOPSIS

> **podman command** [*optional*] *value*

Reviewer: I think I have to clarify that the MANPAGE_SYNTAX is just to get an idea of the syntax. The synopsis was correct in this case.

Author: Although correct, I think it would be much nice to maintain a style! Lmk and I will change it back

Reviewer 2: I think the previous version is better here. You would use commands like `podman --log-level debug ps` so `**podman** [*options*] *command*` is better IMO

**podman** [*options*] *command*
**podman command** [*optional*] *value*

## DESCRIPTION

> Podman (Pod Manager) is a fully featured container engine that is a simple daemonless tool.

Reviewer 3: We want the (Pod Manager) comment here, since we got accused of being a sexist name, and wanted to point out where it comes from.

Author: Haha, but I think the purpose of this page is to describe the command podman not Podman

Reviewer 4: @infinteregrets please retain the "Pod Manager" bit here.

Reviewer 4: And I just read the rest of this.  I'd rather not change this paragraph at all until we can get our @fatherlinux to look at it, but he's on paternity leave through the end of the month.

Author: Okay, I will fix that. After a short description of the command, should the "Pod Manager" part be put in another paragraph? Or should I just revert it back to the way it was in the first place

Reviewer 4: Yes please revert everything here up to the OPTIONS section.  We can make an issue if you'd like to rework this, but I'd really like to get @fatherlinux 's input as he worked up the original as is on paterinity leave this month.

Author: Sure! I have updated the description to the way it was before. Once this is merged, I will open an issue to possibly propose new changes

> The podman command is used to control the Podman container engine using a CLI.  Most commands can be run as a regular user, without requiring additional privileges. Default settings for flags are defined in `containers.conf` which can be found in `/etc/containers`. Most settings for remote connections use the server’s `containers.conf`, except when documented in man pages.

Reviewer 3: containers.conf is installed in /usr/share/containers/containers.conf not in /etc/containers.

Author: On my podman installation:
```
root@ae8790df7bd5:/usr/share/containers# ls
seccomp.json
```
```
root@ae8790df7bd5:/etc/containers# ls
containers.conf  policy.json  registries.conf
```
"Container engines read the /usr/share/containers/containers.conf and /etc/containers/containers.conf files if they exists. When running in rootless mode, they also read $HOME/.config/containers/containers.conf files."
From: https://github.com/containers/common/blob/master/docs/containers.conf.5.md
What would be the correct one?

Reviewer 3: Well that is a packaging issue for your distribution, the way it is supposed to work is 

/usr/share/containers/containers.conf  - Comes from the distribution.
/etc/containers/containers.conf - Is modifications generated by the system admin
$HOME/.config/containers/containers.conf - Is for modifications by the user.

Author: Understood!

> The podman command is used to control the Podman container engine using a CLI.  Most commands can be run as a regular user, without requiring additional privileges. Default settings for flags are defined in `containers.conf` which can be found in `/etc/containers`. Most settings for remote connections use the server’s `containers.conf`, except when documented in man pages.

Reviewer 3: How about

```
The **podman** command is used to manage containers and container images.  Podman can manage containers as a regular user, without requiring additional privileges, as well as run containers as root. Default settings for flags are defined in `containers.conf` which can be found in `/usr/share/containers`.  Podman can also run in remote mode using the `--remove` flag, which communicates over ssh to a remove podman socket activated service.  Settings for remote connections also use the server’s `containers.conf`.
```

Author: Thanks! This is much better. If you have any other suggestions then lmk!

> The podman command is used to control the Podman container engine using a CLI.  Most commands can be run as a regular user, without requiring additional privileges. Default settings for flags are defined in `containers.conf` which can be found in `/etc/containers`. Most settings for remote connections use the server’s `containers.conf`, except when documented in man pages.

Reviewer: The start of this section should be: 
**podman** is used to ...

I wanted to start every DESCRIPTION section with the command itself.

Author: Do you have any suggestions regarding how this should be phrased?

## OPTIONS

#### **--cgroup-manager**=**_systemd_** | *cgroupfs*

| CGroup Manager| Description|
| ----------- | ----------- |
| systemd (Default)| daemon process for managing services|

> | cgroupfs | pseudo-filesystem for the cgroup interface|

Reviewer 3: For some reason this translates in the man page as a third empty box.

Author: I am not sure, this looks fine on my build
![Screen Shot 2021-06-03 at 12 27 53 PM](https://user-images.githubusercontent.com/65443164/120601144-35e7ee80-c467-11eb-9670-1c81656f4a62.png)

Reviewer 3: Strange, what OS and what version of gomd2man are you using?

Here is what this patch looks like when converted to nroff on Fedora 35.
Check out all of the tables.

[podman.1.txt](https://github.com/containers/podman/files/6590864/podman.1.txt)

Author: I am using a Ubuntu 20.04 VM. 
Checkout this gist: https://gist.github.com/infiniteregrets/4a020243a28988411aaa92a91b010719

Works just fine for me. I also made a few changes that fix the formatting issues

Remote connections use the server's containers.conf, except when documented in
man pages.

## OPTIONS

#### **--cgroup-manager**=**_systemd_** | *cgroupfs*

| CGroup Manager| Description|
| ----------- | ----------- |
| **systemd**| daemon process for managing services|
| cgroupfs | pseudo-filesystem for the cgroup interface|

The CGroup manager to use for container cgroups. Default is systemd unless overridden in the `containers.conf` file. The default manager is **systemd**.

**Note:**

- Setting this flag can cause certain commands to break when called on containers previously created by the other CGroup manager type.
- CGroup manager is not supported in rootless mode when using CGroups Version V1.

#### **--cni-config-dir**=**_/etc/cni/net.d_**

Path of the configuration directory for CNI networks.

#### **--connection**, **-c**

Connection to use for remote podman

**Note:**

- Default connection is configured in `containers.conf`.
- Remote connections use local `containers.conf` for default.

#### **--conmon**

Path of the conmon binary.

**Note:**

> - Default path is configured in `containers.conf`

Reviewer 2: Why use a list for only one item?

Podman and libpod currently support an additional `precreate` state which is called before the runtime's `create` operation.  Unlike the other stages, which receive the container state on their standard input, `precreate` hooks receive the proposed runtime configuration on their standard input.  They may alter that configuration as they see fit, and write the altered form to their standard output.

**Warning:**

> - The `precreate` hook allows additional mounts to the runtime configuration.  This might make it easy to break things.  Before reporting libpod errors, try running your container with `precreate` hooks disabled to see if the problem is due to one of your hooks.

Reviewer 4: ```suggestion
- The `precreate` hook allows additional mounts to the runtime configuration.  This might make it easy to break things.  Before reporting libpod errors, try running the container with `precreate` hooks disabled to see if the problem is due to one of the hooks.
```

Podman prompts for the login password on the remote server.

Identity value resolution precedence:
 - command line value.
 - environment variable `CONTAINER_SSHKEY`, if `CONTAINER_HOST` is found.
 - `containers.conf`.
Remote connections use local `containers.conf` for default.

#### **--log-level**=***_error_** | debug | info | warn | fatal/panic*

Log messages above a specified level.

| Argument    | Description |
| ----------- | ----------- |

> | Error (Default) | Runtime/Unexpected error messages.|

Reviewer: Error should just be **bold**, without the DEFAULT behind it.

Author: Fixed, please take another look for other tables as well

Reviewer 3: Warn is now default.

Author: Updated

**Note:**
- Do not pass the leading `--` to the flag. To pass the runc flag `--log-format json`
to podman build, the option given would be `--runtime-flag log-format=json`.

#### **--storage-driver**=*value*

| UID      | Value (Default)|
| ----------- | ----------- |
| 0       | `/etc/containers/storage.conf` |
| Other   | `$HOME/.config/containers/storage.conf`|

Storage driver. 

**Note:**
- Default is *vfs* for non-root users when *fuse-overlayfs* is not available.  

- The `STORAGE_DRIVER` environment variable overrides the default.  The `--storage-driver` specified driver overrides all.

- Overriding this option will cause the *storage-opt* settings in `/etc/containers/storage.conf` to be ignored.  The user must
specify additional options via the `--storage-opt` flag.

#### **--storage-opt**=*value*

| User      | Value (Default)|
| ----------- | ----------- |
| Root        | `/etc/containers/storage.conf`|
| Rootless    | `$HOME/.config/containers/storage.conf`|

Storage driver option.

The `STORAGE_OPTS` environment variable overrides the default. The --storage-opt specified options overrides all. If you specify `--storage-opt=""`, no storage options will be used.

> #### **--syslog**=*true| **false***

Reviewer: The default value should be in the first place.

| Value      | Description |
| ----------- | ----------- |
| True      | Information is logged |
| False   | Information is not logged |

Output logging information to syslog as well as the console.

On remote clients, logging is directed to the file `$HOME/.config/containers/podman.log`.

#### **--tmpdir**

Path to the tmp directory, for libpod runtime content.

**Note:** 
- --tmpdir is not used for the temporary storage of downloaded images.  Use the environment variable `TMPDIR` to change the temporary storage location of downloaded container images. Podman defaults to use `/var/tmp`.

#### **--version**, **-v**

Print the version.

## Environment Variables

Podman can set up environment variables from env of [engine] table in `containers.conf`. These variables can be overridden by passing  environment variables before the `podman` commands.

## Remote Access

> The Podman command can be used with remote services using the `--remote` flag. Connections can

Reviewer: The format of the OPTIONS should stay the same throughout the whole document. In this case **--remote** should be **bold**. 
The term `flag` should be replaced with the term OPTION.

Podman prompts for the login password on the remote server.

Identity value resolution precedence:
 - command line value
 - environment variable `CONTAINER_SSHKEY`, if `CONTAINER_HOST` is found
 - `containers.conf`
Remote connections use local containers.conf for default.
 - command line value.
 - environment variable `CONTAINER_SSHKEY`, if `CONTAINER_HOST` is found.
 - `containers.conf`.
Remote connections use local `containers.conf` for default.

#### **--log-level**=*level*
#### **--log-level**=***_error_** | debug | info | warn | fatal/panic*

> Log messages above specified level: debug, info, warn, error (default), fatal or panic (default: "error")

Reviewer 2: This information is actually outdated. The default is now `warn`. Also the `trace` level is now supported which shows even more information than `debug`.

Author: Thanks! I forgot to verify that. I have updated it.

Podman and libpod currently support an additional `precreate` state which is called before the runtime's `create` operation.  Unlike the other stages, which receive the container state on their standard input, `precreate` hooks receive the proposed runtime configuration on their standard input.  They may alter that configuration as they see fit, and write the altered form to their standard output.

**Warning:**

> - The `precreate` hook allows additional mounts to the runtime configuration.  This might make it easy to break things.  Before reporting libpod errors, try running the container with `precreate` hooks disabled to see if the problem is due to one of the hooks.

Reviewer 2: Same here, why use a list for only one item?

Podman prompts for the login password on the remote server.

Identity value resolution precedence:

- command line value.
- environment variable `CONTAINER_SSHKEY`, if `CONTAINER_HOST` is found.
- `containers.conf`.
Remote connections use local `containers.conf` for default.

> #### **--log-level**=**_error_** | *debug* | *info* | *warn* | *fatal/panic*

Reviewer 2: `fatal` and `panic` are two different log levels so they should be separated like the other options.

Reviewer: From MANPAGE_SYNTAX:
*If the number of arguments is greater than three, the arguments are **not** listed after the equal sign. The arguments have to be shown in a table [...]*

I believe this is the better way here or the command itself will be a bit overloaded.

## CONFIGURATION FILES

#### containers.conf

**Path(s):**

- `/usr/share/containers/containers.conf`
- `/etc/containers/containers.conf`
- `$HOME/.config/containers/containers.conf`

**Info:**

- Podman has builtin defaults for command line options. These defaults can be overridden using the `containers.conf` configuration files.

> - Distributions ship the `/usr/share/containers/containers.conf` file with their default settings. Administrators can override fields in this file by creating the `/etc/containers/containers.conf` file.  Users can further modify defaults by creating the `$HOME/.config/containers/containers.conf` file.

Reviewer 4: ```suggestion
- Distributions ship the `/usr/share/containers/containers.conf` file with their default settings. Administrators can override fields in this file by creating the `/etc/containers/containers.conf` file.  Rootless users can further modify their own defaults by creating the `$HOME/.config/containers/containers.conf` file.
```

- Podman merges its builtin defaults with the specified fields from these files, if they exist. Fields specified in the users file override the administrator's file, which overrides the distribution's file, which override the built-in defaults.

> - Podman uses builtin defaults if no `containers.conf` file is found.

Reviewer 4: ```suggestion
- Podman uses built-in defaults if there is no `containers.conf` file is found.
```

Author: Would you like it to be, "if there is no `containers.conf` file." ?

Images are pulled under `XDG_DATA_HOME` when specified, otherwise in the home directory of the user under `.local/share/containers/storage`.

> Currently the `slirp4netns` package is required to be installed to create a network device, otherwise rootless containers need to run in the network namespace of the host.

Reviewer 4: @mheon does this still hold for 3.2?

Reviewer 5: Yes, this will likely always be true, we've just added the additional requirement of CNI on the host for these installations.

Author: Thanks for the update, are there any changes that should be made to this section that you can think of?

Author: I am leaving this as is

### **NOTE:** Unsupported file systems in rootless mode

> The `Overlay file system (OverlayFS)` is not supported in rootless mode.  The fuse-overlayfs package is a tool that provides the functionality of OverlayFS in user namespace that allows mounting file systems in rootless environments.  It is recommended to install the fuse-overlayfs package.  In rootless mode Podman will automatically use the `fuse-overlafs` program as the `mount_program` if installed, as long as the `$HOME/.config/containers/storage.conf` file was not previously created.  If `storage.conf` exists in the homedir, add `mount_program = "/usr/bin/fuse-overlayfs"` under `[storage.options.overlay]` to enable this feature.

Reviewer 4: ```suggestion
The `Overlay file system (OverlayFS)` is not supported in rootless mode.  The fuse-overlayfs package is a tool that provides the functionality of OverlayFS in user namespace that allows mounting file systems in rootless environments.  It is recommended to install the fuse-overlayfs package.  In rootless mode Podman will automatically use the `fuse-overlayfs` program as the `mount_program` if installed, as long as the `$HOME/.config/containers/storage.conf` file was not previously created.  If `storage.conf` exists in the homedir, add `mount_program = "/usr/bin/fuse-overlayfs"` under `[storage.options.overlay]` to enable this feature.
```
